sklearn/tree/_tree.pyx

```diff
@@ -255,7 +274,36 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
                 # inspection and interpretation
                 splitter.node_value(tree.value + node_id * tree.value_stride)
 
+                middle_value = tree._get_middle_value(node_id)
+
                 if not is_leaf:
+                    if splitter.monotonic_cst[split.feature] == 0:
+                        # Split on a feature with no monotonicity constraint
+                        # Current bounds must always be propagated to both children
+                        # Splitting criterion will always check them in check_monotonicity
+                        left_child_min = lower_bound
+                        left_child_max = upper_bound
+                        right_child_min = lower_bound
+                        right_child_max = upper_bound
+                    elif splitter.monotonic_cst[split.feature] == 1:
+                        # Split on a feature with increasing monotonicity constraint
+                        # Lower bound for left child is the current lower bound
+                        left_child_min = lower_bound
+                        # Upper bound for right child is the current upper bound
+                        right_child_max = upper_bound
+                        # Lower bound for right child and upper bound for left child are set to the same value
+                        right_child_min = middle_value
+                        left_child_max = middle_value
+                    else:  # i.e. splitter.monotonic_cst[split.feature] == -1
+                        # Split on a feature with decreasing monotonicity constraint
+                        # Lower bound for right child is the current lower bound
+                        right_child_min = lower_bound
+                        # Upper bound for left child is the current upper bound
+                        left_child_max = upper_bound
+                        # Lower bound for left child and upper bound for right child are set to the same value
+                        left_child_min = middle_value
+                        right_child_max = middle_value
```

Comment:
This is repeated twice. Maybe worth a short inline function.

Reply:
Since we have 4 output values and this is Cython, not Python, this would mean writing into pointers passed as inputs to the function... not sure what is best in terms of readability.